### PR TITLE
Register a callback that tries to load other versions Open LDAP library if the default is failed

### LINKED ISSUE #69456: System.DirectoryServices.Protocols.LdapConnection fails to find the new version of the LDAP library on Ubuntu 22.04
### Description

Ubuntu 20.04 has LDAP library libldap-2.4 https://packages.ubuntu.com/source/focal/openldap
Ubuntu 22.04 has LDAP library libldap-2.5.11 https://packages.ubuntu.com/source/jammy/openldap

When running .NET Core 6 code on Ubuntu 22.04, System.DirectoryServices.Protocols.LdapConnection cannot find the latest version of LDAP an fails.

### Reproduction Steps

Install the LDAP libraries on Ubuntu 22.04:
```
ubuntu@w-ubuntu-2:~$ sudo apt-get install --no-install-recommends -y krb5-user ldap-utils libsasl2-modules-gssapi-mit wget
[sudo] password for ubuntu:
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
ldap-utils is already the newest version (2.5.11+dfsg-1~exp1ubuntu3).
libsasl2-modules-gssapi-mit is already the newest version (2.1.27+dfsg2-3ubuntu1).
wget is already the newest version (1.21.2-2ubuntu1).
krb5-user is already the newest version (1.19.2-2).
The following package was automatically installed and is no longer required:
  libfreetype6
Use 'sudo apt autoremove' to remove it.
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
ubuntu@w-ubuntu-2:~$
```

Steps to reproduce:
Run the following code from an ASP.NET Core 6 application on Ubuntu 22.04:
```
using System.DirectoryServices.Protocols;
var id = new LdapDirectoryIdentifier(servers[i], true, false);
```

### Expected behavior

System.DirectoryServices.Protocols.LdapConnection finds the 2.5.11 version of LDAP library on Ubuntu 22.04 and connects to the LDAP server.

### Actual behavior

```
System.TypeInitializationException
  HResult=0x80131534
  Message=The type initializer for 'Ldap' threw an exception.
  Source=System.DirectoryServices.Protocols
  StackTrace:
   at Interop.Ldap.ldap_initialize(IntPtr& ld, String uri)
   at System.DirectoryServices.Protocols.ConnectionHandle..ctor(String uri)
   at System.DirectoryServices.Protocols.LdapConnection.InternalInitConnectionHandle(String hostname)
   at System.DirectoryServices.Protocols.LdapConnection.Init()
   at System.DirectoryServices.Protocols.LdapConnection..ctor(LdapDirectoryIdentifier identifier, NetworkCredential credential, AuthType authType)
   at TheLdapService.Server.Services.LdapService.GetLdapConnection() in C:\Users\ubuntu\source\repos\TheLdapService\src\TheLdapService.Server\Services\LdapService.cs:line 58
   at TheLdapService.Server.Services.LdapService.Authenticate(String username, String password) in C:\Users\ubuntu\source\repos\TheLdapService\src\TheLdapService.Server\Services\LdapService.cs:line 199
   at TheLdapService.Server.Services.AuthorizationService.<>c__DisplayClass4_0.<Login>b__0() in C:\Users\ubuntu\source\repos\TheLdapService\src\TheLdapService.Server\Services\AuthorizationService.cs:line 39
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)

  This exception was originally thrown at this call stack:
    [External Code]

Inner Exception 1:
DllNotFoundException: Unable to load shared library 'libldap-2.4.so.2' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: liblibldap-2.4.so.2: cannot open shared object file: No such file or directory


```

### Regression?

Same code worked under Ubuntu 20.04, which has libldap-2.4, the library System.DirectoryServices.Protocols.LdapConnection is looking for.

### Known Workarounds

On Ubuntu 22.04, create a symbolic link from ldap-2.4 to ldap-2.5:
```
ubuntu@w-ubuntu-2:~$ sudo ln -s /usr/lib/x86_64-linux-gnu/libldap-2.5.so.0 /usr/lib/x86_64-linux-gnu/libldap-2.4.so.2

ubuntu@w-ubuntu-2:/usr/lib/x86_64-linux-gnu$ ll | grep ldap
lrwxrwxrwx  1 root root       42 May 17 15:29 libldap-2.4.so.2 -> /usr/lib/x86_64-linux-gnu/libldap-2.5.so.0
lrwxrwxrwx  1 root root       20 Feb 16 12:15 libldap-2.5.so.0 -> libldap-2.5.so.0.1.6
-rw-r--r--  1 root root   376576 Feb 16 12:15 libldap-2.5.so.0.1.6
```

### Configuration

```
ubuntu@w-ubuntu-2:~$ dotnet --info
.NET SDK (reflecting any global.json):
 Version:   6.0.300
 Commit:    8473146e7d

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  22.04
 OS Platform: Linux
 RID:         ubuntu.22.04-x64
 Base Path:   /usr/share/dotnet/sdk/6.0.300/

Host (useful for support):
  Version: 6.0.5
  Commit:  70ae3df4a6

.NET SDKs installed:
  6.0.300 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.5 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.5 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
```

### Other information

_No response_

### PR DESCRIPTION
**Pull request by buyaa-n**
### Register a callback that tries to load other versions of Open LDAP library if the default is failed

Open LDAP library version is hardcoded in runtime and used for native calls
https://github.com/dotnet/runtime/blob/b12e288a143c2986f81aa7c025beac9f5a4c7c29/src/libraries/Common/src/Interop/Linux/Interop.Libraries.cs#L9
From Ubuntu 22.04 only the new version of the library (libldap-2.5.so.0) is included in OS, so [System.DirectoryServices.Protocols.LdapConnection fails to find the new version of the LDAP library](https://github.com/dotnet/runtime/issues/69456) as mentioned in the issue.

@AaronRobinsonMSFT pointed out that we can register a callback to AssemblyLoadContext.Default.ResolvingUnmanagedll event. This event is raised if the native library could not be resolved via the default resolution logic.

The PR doesn't include a specific test, I don't think I can test the logic added here specifically, let me know if you have a testing idea. Though all existing tests that connects to LDAP server would test this, run the tests locally on Ubuntu 22.04, after the fix all tests that require LDAP server passed. (No CI test runs LDAP server so they are skipped in CI)

Fixes https://github.com/dotnet/runtime/issues/69456

### CODE REVIEW

File: src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs
Comment (buyaa-n) [inline src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:82]: Open LDAP version 2.6 is in preview, not sure if the library name would be like this, need to check

File: src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs
Comment (buyaa-n) [inline src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:83]: This kind of redundant, though assume it could be useful for rare mechanical failure case

File: src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs
Comment (buyaa-n) [inline src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:78]: Not sure if we should throw instead in this case

File: src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:72]: Looking at [the doc](https://learn.microsoft.com/dotnet/core/dependency-loading/loading-unmanaged), there might be a better mechanism for this. I'm wondering if we should consider [`NativeLibrary.SetDllImportResolver(Assembly, DllImportResolver)`](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.nativelibrary.setdllimportresolver#system-runtime-interopservices-nativelibrary-setdllimportresolver(system-reflection-assembly-system-runtime-interopservices-dllimportresolver)). @elinor-fung Do you have guidance on which of the resolving mechansims we should be using in this case? I think the `DllImportResolver` callback is always called so perhaps the approach here is correct, but want to verify.

The gist of the scenario is _most_ of the time the unmanaged binary is present on the machine. However, when it isn't, we would like to try newer versions in the case the system has been updated.

File: src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs
Comment (AaronRobinsonMSFT) [reply src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:78]: We shouldn't throw from this API. The system will provide a "missing DLL" exception for us.

File: src/libraries/System.DirectoryServices.Protocols/src/Resources/Strings.resx
Comment (joperezr) [inline src/libraries/System.DirectoryServices.Protocols/src/Resources/Strings.resx:430]: Super small NIT

```suggestion
    <value>The OpenLDAP library with version {0} was not found.</value>
```

File: src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs
Comment (joperezr) [inline src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:81]: Since 2.5 is the more widely used in newer distros, I wonder if we should let 2.5 be the default (meaning the one we use in the Libraries.OpenLdap name) and then we use this event as a fallback for the legacy (and preview) versions.

**Review by joperezr — COMMENTED** [review]
Thanks for doing this @buyaa-n. Changes LGTM but I'll defer to @AaronRobinsonMSFT and @elinor-fung  for what the best approach to hook to that event is.

File: src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs
Comment (elinor-fung) [reply src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:72]: `DllImportResolver` gets called first - before even the 'normal' attempt to load. `ResolvingUnmanagedDll` is fired last - only if all other resolution has failed.

Considerations for choosing between them:
1. Should your handler take precedence over other resolution methods or just be a fallback?
    - `DllImportResolver` will be called before the default attempt to load `Libraries.OpenLdap`. You can fully define the order of what you attempt to load (so you can say use `libldap-2.5.so.0` as long as it is found - regardless of whether `libldap-2.4.so.2` is present)
      - If your assembly has a lot of other p/invokes that your handler doesn't care about, this may also mean a lot of extra/unnecessary calls to a handler
    - `ResolvingUnmanagedDll` is only called as a fallback. It will not be called when other forms of resolution succeed.
2. Will there be multiple such handlers?
    - Only one `DllImportResolver` is allowed per `Assembly`. If one is already set for that assembly, `SetDllImportResolver` will fail.
    - `ResolvingUnmanagedDll` is an event, so you can have as many handlers as you want

> The gist of the scenario is most of the time the unmanaged binary is present on the machine. However, when it isn't, we would like to try newer versions in the case the system has been updated.

I think this would indicate only wanting the fallback for (1) - `ResolvingUnmanagedDll`.

Perhaps a larger concern is (2) - because this file is under Common and it has no guarantee that all the assemblies using it do not set a `DllImportResolver`.

In terms of setting a precedent in `libraries/Common/**/Interop/` files for handling this kind of thing, I think `ResolvingUnmanagedDll` would be the better option.

Comment (elinor-fung) [reply src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:72]: Yep, you want to `GetLoadContext` for the current assembly. Otherwise things will still fail if the assembly is in a non-default ALC (and we'll be unnecessarily adding a handler to the defalut ALC).

File: src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs
Comment (elinor-fung) [inline src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:76]: Might also want to scope this handling to `assembly` matching the assembly with this class.

File: src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs
Comment (buyaa-n) [reply src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:72]: > If your assembly has a lot of other p/invokes that your handler doesn't care about, this may also mean a lot of extra/unnecessary calls to a handler

Yes, we have lots of other p/invokes, mostly those are different for each platform, I wouldn't like to handle them all in the handler 

> ResolvingUnmanagedDll is only called as a fallback. It will not be called when other forms of resolution succeed.

I think that is what I need, the issue is only happening for one Linux library, I want only handle that failure, if it happens, don't want to change/touch other libraries on other platforms

> I think this would indicate only wanting the fallback for (1) - ResolvingUnmanagedDll.

Thank you for broad explanation, completely agree

File: src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs
Comment (buyaa-n) [reply src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:78]: Thanks

File: src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs
Comment (buyaa-n) [reply src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:81]: Sounds good

File: src/libraries/Common/src/Interop/Linux/Interop.Libraries.cs
Comment (steveharter) [inline src/libraries/Common/src/Interop/Linux/Interop.Libraries.cs:12]: Would this also pick up `libldap-2.5.so.0.1.6` for example?

**Review by steveharter — APPROVED** [review]
LGTM pending question on patch versions

File: src/libraries/Common/src/Interop/Linux/Interop.Libraries.cs
Comment (buyaa-n) [reply src/libraries/Common/src/Interop/Linux/Interop.Libraries.cs:12]: It looks like `libldap-2.5.so.0` is not the real library it is linked to `libldap-2.5.so.0.1.6` and uses it.

File: src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs
Comment (elinor-fung) [inline src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:91]: This should just return `IntPtr.Zero`. The runtime will handle throwing `DllNotFoundException` with the library name.

Comment (buyaa-n) [reply src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs]: Thank you, fixed

File: src/libraries/Common/src/Interop/Linux/Interop.Libraries.cs
Comment (AaronRobinsonMSFT) [reply src/libraries/Common/src/Interop/Linux/Interop.Libraries.cs:12]: > It looks like libldap-2.5.so.0 is not the real library

It is likely a symbolic link to the specific version. It is a common mechanism on *nix for handling library versions.

File: src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:84]:
```suggestion
                if (NativeLibrary.TryLoad("libldap-2.4.so.2", out IntPtr handle) ||
                    NativeLibrary.TryLoad("libldap-2.6.so.0", out handle))
```

Is this the priority order we want? I would assume we want to try the latest first. Unsure the customer preference here, but I would think it is leaning toward the latest.

Comment (buyaa-n) [reply src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:84]: 2.6 is in preview, so I assume most likely it not included in any Linux distributions yet, but older versions, or even the latest versions of some distributions might still using the 2.4, therefore preferred checking 2.4 first.

Comment (jkoritzinsky) [reply src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:84]: Instead of using a custom handler, would using a DllImportResolver be a better option?

Comment (AaronRobinsonMSFT) [reply src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:84]: That was discussed and rejected at https://github.com/dotnet/runtime/pull/88851#discussion_r1262908436.

Comment (AaronRobinsonMSFT) [reply src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs:84]: > therefore preferred checking 2.4 first

The issue with this is users will then be required to either (a) remove 2.4 from their system or (b) update this BCL code path to try the preview version. I don't know how common having 2.4 and 2.5 on the same machine is, but if it is common then it makes trying out 2.6 difficult. I think we want 2.6 to be higher priority than 2.4.

**Comment by buyaa-n** [comment]
Failures unrelated and known

**Pull request merged**
